oracle.py

```diff
@@ -1,439 +1,280 @@
-'''import statements'''
```

Comment:
This is really hard to review. What has really changed with respect to the original `oracle.py` (in e.g branch `main`)?